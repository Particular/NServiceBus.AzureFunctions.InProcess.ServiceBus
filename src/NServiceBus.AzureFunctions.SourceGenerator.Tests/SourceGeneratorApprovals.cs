#pragma warning disable IDE0079 // Remove unnecessary suppression
#pragma warning disable CS0618 // Disable obsolete warning for NServiceBusEndpointNameAttribute
namespace NServiceBus.AzureFunctions.SourceGenerator.Tests
{
    using System;
    using System.Collections.Generic;
    using System.Collections.Immutable;
    using System.Linq;
    using Microsoft.CodeAnalysis;
    using Microsoft.CodeAnalysis.CSharp;
    using NUnit.Framework;
    using Particular.Approvals;

    [TestFixture]
    public class SourceGeneratorApprovals
    {
        [Test]
        public void UsingNamespace()
        {
            var source =
@"using NServiceBus;

[assembly: NServiceBusEndpointName(Foo.Startup.EndpointName)]

namespace Foo
{
    public class Startup
    {
        public const string EndpointName = ""endpoint"";
    }
}";
            var (output, _) = GetGeneratedOutput(source);

            Approver.Verify(output);
        }

        [Test]
        public void UsingFullyQualifiedAttributeName()
        {
            var source =
@"[assembly: NServiceBus.NServiceBusEndpointName(Foo.Startup.EndpointName)]

namespace Foo
{
    public class Startup
    {
        public const string EndpointName = ""endpoint"";
    }
}";
            var (output, _) = GetGeneratedOutput(source);

            Approver.Verify(output);
        }

        [Test]
        public void NameIsStringValue()
        {
            var source = @"[assembly: NServiceBus.NServiceBusEndpointName(""endpoint"")]";
            var (output, _) = GetGeneratedOutput(source);

            Approver.Verify(output);
        }

        [Test]
        public void No_attribute_should_not_generate_trigger_function()
        {
            var source = @"";
            var (output, _) = GetGeneratedOutput(source);

            Approver.Verify(output);
        }

        [Test]
        public void No_attribute_should_not_generate_compilation_error()
        {
            var source = @"using NServiceBus;";
            var (output, diagnostics) = GetGeneratedOutput(source);

            Assert.False(diagnostics.Any(d => d.Severity == DiagnosticSeverity.Error));
        }

        [Test]
        public void Can_override_trigger_function_name()
        {
            var source =
                @"using NServiceBus;

[assembly: NServiceBusEndpointName(""endpoint"", ""trigger"")]

public class Startup
{
}";
            var (output, _) = GetGeneratedOutput(source);

            Approver.Verify(output);
        }

        [TestCase(null)]
        [TestCase("")]
        [TestCase(" ")]
        public void Invalid_name_should_cause_an_error(string endpointName)
        {
            var source = @"
using NServiceBus;

[assembly: NServiceBusEndpointName(""" + endpointName + @""")]
";
            var (output, diagnostics) = GetGeneratedOutput(source, suppressGeneratedDiagnosticsErrors: true);

            Assert.True(diagnostics.Any(d => d.Severity == DiagnosticSeverity.Error && d.Id == TriggerFunctionGenerator.InvalidEndpointNameError.Id));
        }

        [TestCase(null)]
        [TestCase("")]
        [TestCase(" ")]
        public void Invalid_trigger_function_name_should_cause_an_error(string triggerFunctionName)
        {
            var source = @"
using NServiceBus;

[assembly: NServiceBusEndpointName(""endpoint"", """ + triggerFunctionName + @""")]
";
            var (output, diagnostics) = GetGeneratedOutput(source, suppressGeneratedDiagnosticsErrors: true);

            Assert.True(diagnostics.Any(d => d.Severity == DiagnosticSeverity.Error && d.Id == TriggerFunctionGenerator.InvalidTriggerFunctionNameError.Id));
        }

        [OneTimeSetUp]
        public void Init()
        {
            // For the unit tests to work, the compilation used by the source generator needs to know that NServiceBusEndpointName
            // is an attribute from NServiceBus namespace and its full name is NServiceBus.NServiceBusEndpointNameAttribute.
            // By referencing NServiceBusEndpointNameAttribute here, NServiceBus.AzureFunctions.InProcess.ServiceBus is forced to load and participate in the compilation.
            _ = new NServiceBusEndpointNameAttribute("test");
            _ = new NServiceBusEndpointNameAttribute("test", "trigger");
        }

        static (string output, ImmutableArray<Diagnostic> diagnostics) GetGeneratedOutput(string source, bool suppressGeneratedDiagnosticsErrors = false)
        {
            var syntaxTree = CSharpSyntaxTree.ParseText(source);
            var references = new List<MetadataReference>();
            var assemblies = AppDomain.CurrentDomain.GetAssemblies();

            foreach (var assembly in assemblies)
            {
                if (!assembly.IsDynamic && !string.IsNullOrWhiteSpace(assembly.Location))
                {
                    references.Add(MetadataReference.CreateFromFile(assembly.Location));
                }
            }

            var compilation = CSharpCompilation.Create("foo", new[] { syntaxTree }, references, new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));

            // Fail tests when the injected program isn't valid _before_ running generators
            var compileDiagnostics = compilation.GetDiagnostics();
            Assert.False(compileDiagnostics.Any(d => d.Severity == DiagnosticSeverity.Error), "Failed: " + compileDiagnostics.FirstOrDefault()?.GetMessage());

            var generator = new TriggerFunctionGenerator();

            var driver = CSharpGeneratorDriver.Create(generator);
            driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out var generateDiagnostics);

            if (!suppressGeneratedDiagnosticsErrors)
            {
                Assert.False(generateDiagnostics.Any(d => d.Severity == DiagnosticSeverity.Error), "Failed: " + generateDiagnostics.FirstOrDefault()?.GetMessage());
            }

            return (outputCompilation.SyntaxTrees.Last().ToString(), generateDiagnostics);
        }
    }
}
#pragma warning restore CS0618
#pragma warning restore IDE0079 // Remove unnecessary suppression