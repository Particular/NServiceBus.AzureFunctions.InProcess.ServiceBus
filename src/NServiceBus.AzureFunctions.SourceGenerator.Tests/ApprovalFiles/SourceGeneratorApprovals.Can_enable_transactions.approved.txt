// <autogenerated/>
using ExecutionContext = Microsoft.Azure.WebJobs.ExecutionContext;
using Microsoft.Azure.ServiceBus;
using Microsoft.Azure.ServiceBus.Core;
using Microsoft.Azure.WebJobs;
using Microsoft.Extensions.Logging;
using System.Threading;
using System.Threading.Tasks;
using NServiceBus;

class FunctionEndpointTrigger
{
    readonly FunctionEndpoint endpoint;

    public FunctionEndpointTrigger(FunctionEndpoint endpoint)
    {
        this.endpoint = endpoint;
    }

    [FunctionName("NServiceBusFunctionEndpointTrigger-endpoint")]
    public async Task Run(
        [ServiceBusTrigger(queueName: "endpoint", AutoComplete = false)]
        Message message,
        MessageReceiver messageReceiver,
        ILogger logger,
        ExecutionContext executionContext,
        CancellationToken cancellationToken)
    {
        await endpoint.ProcessTransactional(message, executionContext, messageReceiver, logger, cancellationToken);
    }
}